chain/chain/src/chain.rs

```diff
@@ -1272,31 +1273,60 @@ impl Chain {
         &mut self,
         shard_id: ShardId,
         sync_hash: CryptoHash,
-        part: ShardStateSyncResponsePart,
+        part_id: u64,
+        num_parts: u64,
+        data: &Vec<u8>,
     ) -> Result<(), Error> {
         let shard_state_header = self.get_received_state_header(shard_id, sync_hash)?;
         let ShardStateSyncResponseHeader { chunk, .. } = shard_state_header;
-        let state_root = &chunk.header.inner.prev_state_root;
+        let state_root = chunk.header.inner.prev_state_root;
         self.runtime_adapter
-            .accept_state_part(state_root, &part.state_part, &part.proof)
+            .validate_state_part(&state_root, part_id, num_parts, data)
             .map_err(|_| ErrorKind::InvalidStatePayload)?;
+
+        // Saving the part data.
+        let mut store_update = self.store.owned_store().store_update();
+        let key = StatePartKey(part_id, state_root).try_to_vec()?;
+        store_update.set_ser(COL_STATE_PARTS, &key, data)?;
+        store_update.commit()?;
         Ok(())
     }
 
     pub fn set_state_finalize(
         &mut self,
         shard_id: ShardId,
         sync_hash: CryptoHash,
+        num_parts: u64,
     ) -> Result<(), Error> {
         let shard_state_header = self.get_received_state_header(shard_id, sync_hash)?;
+        let mut height = shard_state_header.chunk.header.height_included;
+        let state_root = shard_state_header.chunk.header.inner.prev_state_root.clone();
+        let mut parts = vec![];
+        for i in 0..num_parts {
+            let key = StatePartKey(i, state_root.clone()).try_to_vec()?;
+            parts.push(self.store.owned_store().get_ser(COL_STATE_PARTS, &key)?.unwrap());
```

Comment:
somewhere we should potentially remove these parts from storage

Reply:
++